docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+a well-known robust database engine that can support such a load.
+
+Also, a database that can support ACID transactions is important to provide more guarantees that
+the data was successfully inserted in the database and that an acknowledgement can be sent back to the Ingest Service to notify the
+full node to prune the inserted data.
+
+#### [Database Schema](#database-schema)
+
+One of the challenges when implementing this solution is how to design a database schema that can be suitable to return responses that are
+equivalent to the existing endpoints but at the same time offers flexibility in returning customized responses in the future.
+
+Currently, CometBFT uses an abstraction layer from [cometbft-db](https://github.com/cometbft/cometbft-db) in order to support multiple embedded databases. These databases store
+data as key-value pairs using a byte array datatype, for example to set a value for a key:
+
+```go
+func (db *GoLevelDB) Set(key []byte, value []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	if value == nil {
+		return errValueNil
+	}
+	if err := db.db.Put(key, value, nil); err != nil {
+		return err
+	}
+	return nil
+}
+```
+
+Since the RPC Companion stores the information in a relational database, there are opportunities to better structure and normalize the data. Here is the schema definition for a table to persist a normalized `ResultBlock` data structure in the database (Postgresql)
+
+```sql
+-- Table: comet.result_block
+
+CREATE TABLE IF NOT EXISTS comet.result_block
+(
+    block_id_hash bytea NOT NULL,
+    block_id_parts_hash bytea NOT NULL,
+    block_id_parts_total comet.uint32 NOT NULL,
+    block_header_height bigint NOT NULL,
+    block_header_block_time timestamp with time zone NOT NULL,
+    block_header_chain_id text NOT NULL,
+    block_header_version_block comet.uint64 NOT NULL,
+    block_header_version_app comet.uint64 NOT NULL,
+    block_header_data_hash bytea NOT NULL,
+    block_header_last_commit_hash bytea NOT NULL,
+    block_header_validators_hash bytea NOT NULL,
+    block_header_next_validators_hash bytea NOT NULL,
+    block_header_consensus_hash bytea NOT NULL,
+    block_header_app_hash bytea NOT NULL,
+    block_header_last_results_hash bytea NOT NULL,
+    block_header_evidence_hash bytea NOT NULL,
+    block_header_proposer_address bytea NOT NULL,
+    block_header_last_block_id_hash bytea NOT NULL,
+    block_header_last_block_id_parts_hash bytea NOT NULL,
+    block_header_last_block_id_part_total comet.uint32 NOT NULL,
+    block_last_commit_height comet.uint64 NOT NULL,
+    block_last_commit_round comet.uint32 NOT NULL,
+    block_last_commit_block_id_hash bytea NOT NULL,
+    block_last_commit_block_id_parts_total comet.uint32 NOT NULL,
+    block_last_commit_block_id_parts_hash bytea NOT NULL,
+    CONSTRAINT block_pkey PRIMARY KEY (block_header_height),
```

Comment:
Should we make chainID part of the primary key, in case an operator decides to reuse the same DB for more than one chains?

Reply:
that's a good point but in our design we assume one database per chain, so if they want to have a RPC Companion for another chain they would need to store the data in a separate database